lib/smack/DSAWrapper.cpp

```diff
@@ -22,185 +25,70 @@ RegisterPass<DSAWrapper>
-    f1 = f1 || memcpys[i] == n1;
-    f2 = f2 || memcpys[i] == n2;
-  }
-
-  if (!f1)
-    memcpys.push_back(eqs.getLeaderValue(n1));
-  if (!f2)
-    memcpys.push_back(eqs.getLeaderValue(n2));
-}
-
 void DSAWrapper::getAnalysisUsage(llvm::AnalysisUsage &AU) const {
   AU.setPreservesAll();
-  AU.addRequiredTransitive<llvm::BUDataStructures>();
-  AU.addRequiredTransitive<llvm::TDDataStructures>();
-  AU.addRequiredTransitive<llvm::DSNodeEquivs>();
-  AU.addRequired<dsa::TypeSafety<llvm::TDDataStructures>>();
+  AU.addRequiredTransitive<sea_dsa::DsaAnalysis>();
 }
 
 bool DSAWrapper::runOnModule(llvm::Module &M) {
   dataLayout = &M.getDataLayout();
-  TD = &getAnalysis<llvm::TDDataStructures>();
-  BU = &getAnalysis<llvm::BUDataStructures>();
-  nodeEqs = &getAnalysis<llvm::DSNodeEquivs>();
-  TS = &getAnalysis<dsa::TypeSafety<llvm::TDDataStructures>>();
-  memcpys = collectMemcpys(M, new MemcpyCollector(nodeEqs));
-  staticInits = collectStaticInits(M);
+  SD = &getAnalysis<sea_dsa::DsaAnalysis>().getDsaAnalysis();
+  assert(SD->kind() == sea_dsa::GlobalAnalysisKind::CONTEXT_INSENSITIVE &&
+         "Currently we only want the context-insensitive sea-dsa.");
+  DG = &SD->getGraph(*M.begin());
+  collectStaticInits(M);
+  collectMemOpds(M);
+  countGlobalRefs();
   module = &M;
   return false;
 }
 
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectMemcpys(llvm::Module &M, MemcpyCollector *mcc) {
-
-  for (llvm::Module::iterator func = M.begin(), e = M.end(); func != e;
-       ++func) {
-
-    for (llvm::Function::iterator block = func->begin(); block != func->end();
-         ++block) {
-
-      mcc->visit(*block);
-    }
-  }
-  return mcc->getMemcpys();
-}
-
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectStaticInits(llvm::Module &M) {
-  std::vector<const llvm::DSNode *> sis;
+void DSAWrapper::collectStaticInits(llvm::Module &M) {
   for (GlobalVariable &GV : M.globals()) {
     if (GV.hasInitializer()) {
       if (auto *N = getNode(&GV)) {
-        sis.push_back(N);
+        staticInits.insert(N);
       }
     }
   }
-  return sis;
 }
 
-DSGraph *DSAWrapper::getGraphForValue(const Value *V) {
-  if (const Instruction *I = dyn_cast<Instruction>(V))
-    return TD->getDSGraph(*I->getParent()->getParent());
-  else if (const Argument *A = dyn_cast<Argument>(V))
-    return TD->getDSGraph(*A->getParent());
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return TD->getDSGraph(*BB->getParent());
-  else if (isa<GlobalValue>(V))
-    return TD->getGlobalsGraph();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users()) {
-    return getGraphForValue(U);
+void DSAWrapper::collectMemOpds(llvm::Module &M) {
+  for (auto &f : M) {
+    for (inst_iterator I = inst_begin(&f), E = inst_end(&f); I != E; ++I) {
+      if (MemCpyInst *memcpyInst = dyn_cast<MemCpyInst>(&*I)) {
+        memOpds.insert(getNode(memcpyInst->getSource()));
```

Comment:
Can `getNode` here ever be null?
I would guess that undef could flow into memcpy for example. Not sure.
What happens if it is null and null gets added into `memOpds`?

Reply:
I think we don't have to worry about `getNode` being nullptr here because `getNumGlobals` will trigger assertion failure first before `isMemOpd` is called.